nifi-nar-bundles/nifi-deltalake-bundle/nifi-deltalake-processors/src/main/java/org/apache/nifi/processors/deltalake/DeltaLakeMetadataWriter.java

```diff
@@ -0,0 +1,264 @@
+            .name("parquet-structure")
+            .displayName("Parquet structure in json")
+            .description("Describes the data structure of the parquet file in json format")
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .required(true)
+            .build();
+
+    public static final Relationship REL_SUCCESS = new Relationship.Builder()
+            .name("success")
+            .description("DeltaLake table successfully updated")
+            .build();
+
+    public static final Relationship REL_FAILED = new Relationship.Builder()
+            .name("failure")
+            .description("DeltaLake table update failed")
+            .build();
+
+    private DeltaLakeService deltalakeService;
+
+    @Override
+    protected void init(final ProcessorInitializationContext context) {
+        this.relationships = Collections.unmodifiableSet(new HashSet<>(Arrays.asList(REL_SUCCESS, REL_FAILED)));
+        this.descriptors = Collections.unmodifiableList(Arrays.asList(STORAGE_SELECTOR, LOCAL_PATH, STRUCTURE_JSON,
+                S3_ACCESS_KEY, S3_SECRET_KEY, S3_BUCKET, S3_PATH,
+                AZURE_ACCOUNT_KEY, AZURE_STORAGE_NAME, AZURE_STORAGE_ACCOUNT, AZURE_PATH,
+                GCP_ACCOUNT_JSON_KEYFILE_PATH, GCP_BUCKET, GCP_PATH));
+        deltalakeService = new DeltaLakeService();
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return this.relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return descriptors;
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.create();
+        try {
+            Map<String, String> updateResult = updateDeltaLake();
+            session.putAllAttributes(flowFile, updateResult);
+            session.transfer(flowFile, REL_SUCCESS);
+        } catch (Exception e) {
+            session.putAttribute(flowFile, "delta table update failed with error", e.getMessage());
+            session.transfer(flowFile, REL_FAILED);
+            context.yield();
+        }
+    }
+
+    @OnScheduled
+    public void onScheduled(ProcessContext processContext) {
+        deltalakeService.initialize(processContext);
+    }
+
+    private Map<String, String> updateDeltaLake() {
+        Map<String, String> updateResult = new HashMap<>();
+
+        if (deltalakeService.deltaTableExists()) {
+            deltalakeService.startTransaction();
+            int numberOfAddedFiles = deltalakeService.addNewFilesToDeltaTable();
+            int numberOfRemovedFiles = deltalakeService.removeMissingFilesFromDeltaTable();
+            deltalakeService.finishTransaction();
+
+            updateResult.put("number of new files in the Delta table", String.valueOf(numberOfAddedFiles));
```

Comment:
Please use flowfile attribute names in the format of lower case words separated with dot, see e.g. 
https://github.com/apache/nifi/blob/main/nifi-commons/nifi-utils/src/main/java/org/apache/nifi/flowfile/attributes/FragmentAttributes.java. 

The "delta table update failed with error" can be simply "error.message".

Reply:
done, thanks